libimage/push.go

```diff
@@ -24,33 +30,93 @@ type PushOptions struct {
+		image, resolvedSource, err := r.LookupImage(source, nil)
+		if err != nil {
+			return nil, err
+		}
+
+		// Make sure we have a proper destination, and parse it into an image
+		// reference for copying.
+		if destination == "" {
+			// Doing an ID check here is tempting but false positives (due
+			// to a short partial IDs) are more painful than false
+			// negatives.
+			destination = resolvedSource
+		}
+
+		return pushImage(ctx, image, destination, options, resolvedSource, r)
+	}
+
+	// Below handles the AllTags option, for which we have to build a list of
+	// all the local images that match the provided repository and then push them.
+	//
+	// For now, make sure a destination was not specified and get it from the source.
+	// This could change in the future, but that gets close to the Copy() functionality.
+	if len(destination) != 0 {
+		return nil, fmt.Errorf("`destination` should not be specified if using AllTags")
+	}
+
+	// Make sure the source repository does not have a tag
+	srcNamed, err := reference.ParseNormalizedNamed(source)
 	if err != nil {
 		return nil, err
 	}
+	if !reference.IsNameOnly(srcNamed) {
+		return nil, fmt.Errorf("can't push with AllTags if source tag is specified")
+	}
+
+	logrus.Debugf("Finding all images for source %s", srcNamed.Name())
+	listOptions := &ListImagesOptions{}
+	srcImages, _ := r.ListImages(ctx, []string{srcNamed.Name()}, listOptions)
 
+	// Push each tag for every image in the list
+	for _, img := range srcImages {
+		namedTagged, err := img.NamedTaggedRepoTags()
+		if err != nil {
+			return nil, err
+		}
+		for _, n := range namedTagged {
+			// Filter on repo name again to avoid pushing an image that matches
+			// the source image ID but has a different repository than the source
+			currentNamed, err := reference.ParseNormalizedNamed(n.Name())
+			if err != nil {
+				return nil, err
+			}
+			if reference.Path(currentNamed) == reference.Path(srcNamed) {
+				// Have to use Sprintf because pushImage expects a string
+				destWithTag := fmt.Sprintf("%s:%s", source, n.Tag())
+				_, err := pushImage(ctx, img, destWithTag, options, "", r)
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+	}
+
+	return nil, nil
+}
+
+// pushImage sends a single image to be copied to the destination
+func pushImage(ctx context.Context, image *Image, destination string, options *PushOptions, resolvedSource string, r *Runtime) ([]byte, error) {
 	srcRef, err := image.StorageReference()
 	if err != nil {
 		return nil, err
 	}
 
-	// Make sure we have a proper destination, and parse it into an image
-	// reference for copying.
-	if destination == "" {
-		// Doing an ID check here is tempting but false positives (due
-		// to a short partial IDs) are more painful than false
-		// negatives.
-		destination = resolvedSource
-	}
-
-	logrus.Debugf("Pushing image %s to %s", source, destination)
+	logrus.Debugf("Pushing image %s to %s", transports.ImageName(srcRef), destination)
```

Comment:
This would now print a fairly large c/storage reference. I’m honestly completely unsure about this — on one hand the output will be less readable, OTOH actually listing the _precise_ image we are pushing is quite valuable.

So I’m fine with this change, just highlighting this for @vrothberg .